### PR TITLE
feat: adding PersistenceLayer

### PR DESCRIPTION
**Pull request by jeffrey-baker-vg**
### Related issues, RFCs

**Issue number:** 

### PR status

***Is this ready for review?:*** NO  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [ ] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/idempotency/src/persistence/PersistenceLayer.ts [hunk 1]
```diff
@@ -1,21 +1,96 @@
...
   protected abstract _deleteRecord(): Promise<void>;
   protected abstract _getRecord(): Promise<IdempotencyRecord>;
   protected abstract _putRecord(record: IdempotencyRecord): Promise<void>;
-  protected abstract _updateRecord(): Promise<void>;
+  protected abstract _updateRecord(record: IdempotencyRecord): Promise<void>;
+
+  private generateHash(data: string): string{
+    this.hashFunction.update(data);
+    
+    return this.hashFunction.digest('base64');
```
Comment (heidimorey-vg) [inline packages/idempotency/src/persistence/PersistenceLayer.ts]: Any chance this method will be called multiple times for a class instance? If so, I believe this method will throw
https://nodejs.org/api/crypto.html#hashdigestencoding

Comment (jeffrey-baker-vg) [reply packages/idempotency/src/persistence/PersistenceLayer.ts]: You are absolutely correct. I had this one my list of things to fix and then...didn't.

File: packages/idempotency/src/persistence/PersistenceLayer.ts [hunk 2]
```diff
@@ -1,21 +1,96 @@
...
     return Promise.resolve({} as IdempotencyRecord);
   }
-  public async saveInProgress(): Promise<void> { }
+
+  /**
+   * Saves a record indicating that the function's execution is currently in progress
+   * 
+   * @param data - the data payload that will be hashed to create the hash portion of the idempotency key
+   */
+  public async saveInProgress(data: unknown): Promise<void> {
```
Comment (heidimorey-vg) [inline packages/idempotency/src/persistence/PersistenceLayer.ts:65]: Perhaps this is out of scope, but is there a use case where 'data' would be undefined, such as if there is no input to the lambda? If so, are we comfortable with how the hash logic will respond?

File: packages/idempotency/src/persistence/PersistenceLayer.ts [hunk 3]
```diff
@@ -1,24 +1,161 @@
...
+    return this.envVarsService;
+  }
+
+  /**
+   * Creates the expiry timestamp for the idempotency record
+   * 
+   * @returns the expiry time for the record expressed as number of seconds past the UNIX epoch
+   */
+  private getExpiryTimestamp(): number {
+    const currentTime: number = Date.now() / 1000;
```
Comment (vgphoenixcampos) [inline packages/idempotency/src/persistence/PersistenceLayer.ts:130]: We may be out of sync here, the dynamo persistence layer does a comparison against Date.now() to determine whether the idempotency key is expired. So right now we would be making a comparison of milliseconds vs seconds as our conditional write.

```  

const idempotencyKeyExpired = '#expiry < :now';
    const notInProgress = 'NOT #status = :inprogress';
    const conditionalExpression = `${idempotencyKeyDoesNotExist} OR ${idempotencyKeyExpired} OR ${notInProgress}`;
    try {
      await table.put({ TableName: this.tableName, Item: item, ExpressionAttributeNames: { '#id': this.key_attr, '#expiry': this.expiry_attr, '#status': this.status_attr }, ExpressionAttributeValues: { ':now': Date.now(), ':inprogress': IdempotencyRecordStatus.INPROGRESS }, ConditionExpression: conditionalExpression });
    } catch (e){
    
    
    ```

**Review by vgphoenixcampos — CHANGES_REQUESTED** [review]

File: packages/idempotency/tests/unit/EnvironmentVariableService.test.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  describe('Method: getLambdaFunctionName', ()=> {
+    beforeEach(() => {
+      process.env.AWS_LAMBDA_FUNCTION_NAME = 'testFunction';
+    });
+
+    afterEach(()=> {
+      delete process.env.AWS_LAMBDA_FUNCTION_NAME;
+    });
+
+    test('When called it getse the Lambda function name from the environment variable', ()=> {
```
Comment (vgphoenixcampos) [inline packages/idempotency/tests/unit/EnvironmentVariableService.test.ts:18]: small typo

File: packages/idempotency/src/persistence/PersistenceLayer.ts (hunk 3 above)
Comment (jeffrey-baker-vg) [reply packages/idempotency/src/persistence/PersistenceLayer.ts:130]: The reason I did that is that the configuration item is expiresAfterSeconds

File: packages/idempotency/src/persistence/PersistenceLayer.ts [hunk 5]
```diff
@@ -1,24 +1,161 @@
...
+   * @param data - the data payload that will be hashed to create the hash portion of the idempotency key
+   * @param result - the result of the successfully completed function
+   */
+  public async saveSuccess(data: unknown, result: unknown): Promise<void> { 
+    const idempotencyRecord: IdempotencyRecord = 
+    new IdempotencyRecord(this.getHashedIdempotencyKey(data),
+      IdempotencyRecordStatus.COMPLETED,
+      this.getExpiryTimestamp(),
+      undefined,
+      JSON.stringify(result),
```
Comment (vgphoenixcampos) [inline packages/idempotency/src/persistence/PersistenceLayer.ts]: I found this while making a fix for another issue, it seems like our types aren't matching here. The Idempotency Record response is currently defined as an object | undefined, but here you're passing in a string. It looks like we diverged a bit from the Python, where the Python converts the response to a string and puts that into a record, but in our implementation of the record it accepts the direct response object.

**Pull request merged**